doc/developer/design/20240508_subsource_schemas_after_planning.md

```diff
@@ -0,0 +1,113 @@
+    `PurificationError` wraps some other error that expresses the semantics of
+    handling PostgreSQL `TEXT COLUMNS`.
+
+## Out of Scope
+
+-   Actually creating subsources for Kafka sources.
+-   Schema migrations
+-   Fully separating concerns w/r/t purification, i.e. I am not planning to
+    address [MySQL source's blending of
+    concerns](https://github.com/MaterializeInc/materialize/issues/26999),
+    though I think that the spirit of this change should include that task.
+-   Supporting `ALTER SOURCE...ADD SUBSOURCE` for MySQL, though that should be
+    relatively simple to do.
+
+## Solution Proposal
+
+1.  Modularize source planning. Source planning will become more complex and the
+    current code structure stretches many people's ability to comprehend what's
+    happening.
+2.  Move the API for planning that occurs during purification into a spot more
+    semantically appropriate, e.g. let the `SouceConnection` trait define how
+    its options should be allowed to be used. e.g. we should provide common
+    errors that both `PurificationError` (for new sources) and `AdapterError`
+    (for additional subsources) can wrap to signal that there are issues with
+    source definitions.
+3.  Restructure purification such that we track the names of the sources we've
+    been requested to create, as well as the "upstream" objects to which they
+    refer––i.e. instead of generating `Vec<CreateSubsourceStatement>` we should
+    generate `BTreeMap<SubsourceName, UpstreamObjectName>`.
+4.  Add a fn to the `SourceConnection` trait like:
+
+    ```rust
+    fn get_output_schema_for_name(
+        &self,
+        upstream_object_name: UpstreamObjectName
+    ) -> Result<RelationDesc, Err>
+    ```
+
+    This will let us ask for the given schema of a subsource after the source
+    itself is planned.
+
+    `UpstreamObjectName` in the code is likely to just be `UnresolvedItemName`.
+
+5.  After planning the primary source, plan all subsources by generating
+    `CreateSourceStatement`s using the name, reference mapping; we'll get the
+    schema for the new objects by looking them up in the `Source`.
+
+### `ALTER SOURCE...ADD SUBSOURCE`
+
+We should also refactor the way that `ALTER SOURCE...ADD SUBSOURCE` works to
+instead describe the data it wishes add to the source, and then merge it into
+the existing source's "external metadata," e.g. the `Vec<PostgresTableDesc>`,
+obviously error checking for duplicate references, etc.
+
+To support this, we can extend the `SourceConnection` trait further to
+describe the kind of transformations we will allow to each type of source.
+For PG and MySQL this is simple to envision as something like:
+
+```rust
+fn merge_external_data<T, O>(
+    &mut self, metadata: Vec<T>, options: Vec<O>
+) -> Result<(), Err>
+```
+
+Then, being able to add subsources to a source will just require implementing a
+merge algorithm for the new metadata and options.
```

Comment:
I like this idea a lot - will definitely simplify the work in adding new source types going forward and understanding how they work

Reply:
Mentioned elsewhere but I added a "Mind Palace" section that we might be able to just write the initial impls of a source in terms of this, and then every new source supports `ALTER SOURCE...ADD SUBSOURCE` by default.

Reply:
yes that context was helpful, thanks!